File: iOS/ePotato/ePotato/Source/LocalDatabaseHandler.swift
import SQLite

class LocalDatabaseHandler {

    static let UID = "uid"
    static let EPID = "epid"
    static let PASSWORD = "password"
    static let PHONE = "phone"
    static let FIID = "fiid"
    static let OLD_UID = "old_uid"
    static let TOTAL_STRANGERS = "total_strangers"
    static let CREEPS = "creeps"
    static let FEED_NOTIFICATIONS = "feed_notifications"
    static let FOLLOWERS = "followers"

    static let PREMIUM = "premium"
    static let RESTORED = "restored"

    static let ID = "id"
    static let DT = "dt"
    static let CONTACT_NAME = "name"
    static let CONTACT_PHONE = "phone"
    static let POTATO_TEXT = "text"
    static let POTATO_FORM = "form"

    static let CONTACTS_SCORE = "score"
    static let RECEIVED_POTATOES_TS = "ts"
    static let NEW_POTATOES_PID = "pid"
    static let SENT_POTATOES_UIDS = "uids"
    static let SENT_POTATOES_TSS = "tss"
    static let SENT_POTATOES_NAMES = "names"
    static let SENT_POTATOES_CODE = "code"

    let id = Expression<String?>(ID)
    let uid = Expression<String?>(UID)
    let epid = Expression<String?>(EPID)
    let dt = Expression<String?>(DT)
    let contactName = Expression<String?>(CONTACT_NAME)
    let contactPhone = Expression<String?>(CONTACT_PHONE)
    let potatoText = Expression<String?>(POTATO_TEXT)
    let potatoForm = Expression<String?>(POTATO_FORM)

    let contactsScore = Expression<String?>(CONTACTS_SCORE)
    let receivedPotatosTs = Expression<String?>(RECEIVED_POTATOES_TS)
    let newPotatoesPid = Expression<String?>(NEW_POTATOES_PID)
    let sentPotatoesUids = Expression<String?>(SENT_POTATOES_UIDS)
    let sentPotatoesTss = Expression<String?>(SENT_POTATOES_TSS)
    let sentPotatoesNames = Expression<String?>(SENT_POTATOES_NAMES)
    let sentPotatoesCode = Expression<String?>(SENT_POTATOES_CODE)
    static let SENT_POTATOES_CODE_NULL = 0
    static let SENT_POTATOES_CODE_SENDING = 1
    static let SENT_POTATOES_CODE_ERROR = 2
    static let SENT_POTATOES_CODE_SENT = 3
    static let SENT_POTATOES_CODE_RECEIVED = 4

    var columns = [String: Expression<String?>]()

    let contacts = Table("contacts")
    let receivedPotatoes = Table("received_potatoes")
    let sentPotatoes = Table("sent_potatoes")
    let newPotatoes = Table("new_potatoes")
    let phoneContacts = Table("phone_contacts")
    let tempContacts = Table("temp_contacts")
    let suggestedContacts = Table("suggested_contacts")
    let suggestedFollowing = Table("suggested_following")
    let feedPotatoes = Table("feed_potatoes")
    let newFeedPotatoes = Table("new_feed_potatoes")
    let following = Table("following")

    private static let DELETE = "*DELETE*"

    let db: Connection

    init() {
        let path = NSSearchPathForDirectoriesInDomains(
                .documentDirectory, .userDomainMask, true
        ).first!
        db = try! Connection("\(path)/db.sqlite3")
        try! db.run(contacts.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(uid)
            t.column(epid)
            t.column(contactName)
            t.column(contactPhone)
            t.column(contactsScore)
        })
        try! db.run(receivedPotatoes.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(uid)
            t.column(receivedPotatosTs)
            t.column(dt)
            t.column(potatoText)
            t.column(potatoForm)
        })
        try! db.run(sentPotatoes.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(sentPotatoesUids)
            t.column(sentPotatoesTss)
            t.column(sentPotatoesNames)
            t.column(dt)
            t.column(sentPotatoesCode)
            t.column(potatoText)
            t.column(potatoForm)
        })
        try! db.run(newPotatoes.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(newPotatoesPid)
        })
        try! db.run(phoneContacts.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(contactName)
            t.column(contactPhone)
        })
        try! db.run(tempContacts.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(uid)
            t.column(epid)
        })
        try! db.run(suggestedContacts.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(uid)
            t.column(epid)
            t.column(contactsScore)
        })
        try! db.run(suggestedFollowing.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(uid)
            t.column(epid)
            t.column(contactsScore)
        })
        try! db.run(feedPotatoes.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(uid)
            t.column(dt)
            t.column(potatoText)
            t.column(potatoForm)
        })
        try! db.run(newFeedPotatoes.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(newPotatoesPid)
        })
        try! db.run(following.create(ifNotExists: true) { t in
            t.column(id, primaryKey: true)
            t.column(uid)
            t.column(epid)
        })
        columns = [
                LocalDatabaseHandler.ID: id,
                LocalDatabaseHandler.UID: uid,
                LocalDatabaseHandler.EPID: epid,
                LocalDatabaseHandler.DT: dt,
                LocalDatabaseHandler.CONTACT_NAME: contactName,
                LocalDatabaseHandler.CONTACT_PHONE: contactPhone,
                LocalDatabaseHandler.POTATO_TEXT: potatoText,
                LocalDatabaseHandler.POTATO_FORM: potatoForm,
                LocalDatabaseHandler.CONTACTS_SCORE: contactsScore,
                LocalDatabaseHandler.RECEIVED_POTATOES_TS: receivedPotatosTs,
                LocalDatabaseHandler.NEW_POTATOES_PID: newPotatoesPid,
                LocalDatabaseHandler.SENT_POTATOES_UIDS: sentPotatoesUids,
                LocalDatabaseHandler.SENT_POTATOES_TSS: sentPotatoesTss,
                LocalDatabaseHandler.SENT_POTATOES_NAMES: sentPotatoesNames,
                LocalDatabaseHandler.SENT_POTATOES_CODE: sentPotatoesCode
        ]
    }

    func dropEverything() {
        truncate(contacts)
        truncate(receivedPotatoes)
        truncate(sentPotatoes)
        truncate(newPotatoes)
        truncate(phoneContacts)
        truncate(tempContacts)
        truncate(suggestedContacts)
        truncate(suggestedFollowing)
        truncate(feedPotatoes)
        truncate(newFeedPotatoes)
        truncate(following)
    }

    func get(table: Table, idKey: Expression<String?>, idValue: String, column: Expression<String?>) -> String? {
        return try! db.scalar(table.select(column).filter(idKey == idValue))
    }

    func getAll(table: Table, idKey: Expression<String?>, idValue: String, column: Expression<String?>) -> [String?] {
        var result = [String?]()
        for row in try! db.prepare(table.select(column).filter(idKey == idValue)) {
            result.append(row[column])
        }
        return result
    }

    func getAll(table: Table, column: Expression<String?>) -> [String?] {
        var result = [String?]()
        for row in try! db.prepare(table.select(column)) {
            result.append(row[column])
        }
        return result
    }

    func getAll(table: Table, idKey: Expression<String?>, idValue: String) -> [[String: String?]] {
        var result = [[String: String?]]()
        for row in try! db.prepare(table.filter(idKey == idValue)) {
            var data = [String: String?]()
            for column in row.columnNames.values {
                var key = Array(row.columnNames.keys)[column] as String
                key = key.substring(with: key.index(key.startIndex, offsetBy: 1)..<key.index(key.endIndex, offsetBy: -1))
                data[key] = try! row.get(columns[key]!)
            }
            result.append(data)
        }
        return result
    }

    func getAll(table: Table) -> [[String: String?]] {
        var result = [[String: String?]]()
        for row in try! db.prepare(table) {
            var data = [String: String?]()
            for column in row.columnNames.values {
                var key = Array(row.columnNames.keys)[column] as String
                key = key.substring(with: key.index(key.startIndex, offsetBy: 1)..<key.index(key.endIndex, offsetBy: -1))
                data[key] = try! row.get(columns[key]!)
            }
            result.append(data)
        }
        return result
    }

    @discardableResult
    fileprivate func inup(insert: Bool, _update: Bool, table: Table, pairs: [String?]) -> String? {
        var result: String?
        let delete = pairs[0] == LocalDatabaseHandler.DELETE || pairs.count > 2 && pairs[2] == LocalDatabaseHandler.DELETE
        let update = !delete && pairs[0] != nil && pairs[0] != "" && get(table: table, idKey: columns[pairs[0]!]!, idValue: pairs[1]!, column: columns[pairs[0]!]!) != nil && !insert
        if delete {
            try! db.run((pairs[0] == LocalDatabaseHandler.DELETE ? table : table.filter(columns[pairs[0]!]! == pairs[1]!)).delete())
        } else if update {
            var set = [Setter]()
            for i in 2..<pairs.count {
                if i % 2 == 0 {
                    set.append(columns[pairs[i]!]! <- pairs[i + 1])
                }
            }
            if pairs.count > 2 {
                try! db.run(table.filter(columns[pairs[0]!]! == pairs[1]!).update(set))
            }
        } else if !_update {
            var set = [Setter]()
            if pairs[0] != LocalDatabaseHandler.ID {
                let B64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
                var id: String
                repeat {
                    id = ""
                    for _ in 1...8 {
                        id += String(B64[B64.index(B64.startIndex, offsetBy: Int(arc4random_uniform(UInt32(B64.characters.count))))])
                    }
                } while (getAll(table: table, column: columns[LocalDatabaseHandler.ID]!).map {
                    $0 ?? ""
                }).contains(id)
                set.append(self.id <- id)
                result = id
            } else {
                result = pairs[1]
            }
            for i in 0..<pairs.count {
                if i % 2 == 0 {
                    set.append(columns[pairs[i]!]! <- pairs[i + 1])
                }
            }
            do {
                try db.run(table.insert(set))
            } catch {
                result = nil
            }
        }
        return result
    }

    @discardableResult
    func inup(table: Table, pairs: [String?]) -> String? {
        return inup(insert: false, _update: false, table: table, pairs: pairs)
    }

    @discardableResult
    func insert(table: Table, pairs: [String?]) -> String? {
        return inup(insert: true, _update: false, table: table, pairs: pairs)
    }

    func update(table: Table, pairs: [String?]) {
        inup(insert: false, _update: true, table: table, pairs: pairs)
    }

    func delete(table: Table, idKey: String, idValue: String) {
        inup(insert: false, _update: false, table: table, pairs: [idKey, idValue, LocalDatabaseHandler.DELETE])
    }

    func truncate(_ table: Table) {
        try! db.run(table.delete())
    }

    static func getTimestamp() -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyyMMddHHmmss"
        return dateFormatter.string(from: Date())
    }

    static func getNiceDate(_ timestamp: String) -> String {
        let thenDF = DateFormatter()
        thenDF.dateFormat = "yyyyMMddHHmmss"
        let then = thenDF.date(from: timestamp)!
        let formatter = DateFormatter()
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .short
        if dateFormatter.string(from: then) == dateFormatter.string(for: Date()) {
            formatter.timeStyle = .short
        } else {
            formatter.dateStyle = .short
        }
        return formatter.string(from: then)
    }

    static func explode(_ input: String?) -> [String] {
        return input != nil ? input!.components(separatedBy: ",") : [""]
    }

    static func implode(_ input: [String]) -> String {
        var result = input[0]
        for x in input[1..<input.count] {
            result += ",\(x)"
        }
        return result
    }

}

File: Android, API/ePotato/app/src/main/java/com/onnoeberhard/potato/LocalDatabaseHandler.java
package com.onnoeberhard.potato;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.support.annotation.Nullable;

import org.apache.commons.codec.DecoderException;
import org.apache.commons.codec.binary.Hex;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TimeZone;

class LocalDatabaseHandler extends SQLiteOpenHelper {

    static final String UID = "uid";
    static final String EPID = "epid";
    static final String PASSWORD = "password";
    static final String PHONE = "phone";
    static final String FIID = "fiid";
    static final String PERMISSION_CONTACTS = "permission_contacts";
    static final String FIRST_PHONE_CONTACTS_JOB = "first_phone_contacts_job";
    static final String CONTACTS_PERMISSION_TRY = "contacts_permission_try";
    static final String OLD_UID = "old_uid";
    static final String TOTAL_STRANGERS = "total_strangers";
    static final String CREEPS = "creeps";
    static final String FEED_NOTIFICATIONS = "feed_notifications";
    static final String FOLLOWERS = "followers";

    static final String PREMIUM = "premium";
    static final String DONATION = "don";
    static final String HI_SCORE = "hi_score";

    static final String ID = "id";
    static final String DT = "dt";
    static final String CONTACT_NAME = "name";
    static final String CONTACT_PHONE = "phone";
    static final String POTATO_TEXT = "text";
    static final String POTATO_FORM = "form";

    static final String CONTACTS = "contacts";
    static final String CONTACTS_SCORE = "score";
    static final String RECEIVED_POTATOES = "received_potatoes";
    static final String RECEIVED_POTATOES_TS = "ts";
    static final String SENT_POTATOES = "sent_potatoes";
    static final String SENT_POTATOES_UIDS = "uids";
    static final String SENT_POTATOES_TSS = "tss";
    static final String SENT_POTATOES_NAMES = "names";
    static final String SENT_POTATOES_CODE = "code";
    static final int SENT_POTATOES_CODE_NULL = 0;
    static final int SENT_POTATOES_CODE_SENDING = 1;
    static final int SENT_POTATOES_CODE_ERROR = 2;
    static final int SENT_POTATOES_CODE_SENT = 3;
    static final int SENT_POTATOES_CODE_RECEIVED = 4;
    static final String NEW_POTATOES = "new_potatoes";
    static final String NEW_POTATOES_PID = "pid";
    static final String PHONE_CONTACTS = "phone_contacts";
    static final String TEMP_CONTACTS = "temp_contacts";
    static final String SUGGESTED_CONTACTS = "suggested_contacts";
    static final String SUGGESTED_FOLLOWING = "suggested_following";
    static final String FEED_POTATOES = "feed_potatoes";
    static final String NEW_FEED_POTATOES = "new_feed_potatoes";
    static final String FOLLOWING = "following";

    private static final String EPOTATO = "epotato";
    private static final String[] TABLES = {CONTACTS, RECEIVED_POTATOES, SENT_POTATOES, NEW_POTATOES, PHONE_CONTACTS, TEMP_CONTACTS, SUGGESTED_CONTACTS, SUGGESTED_FOLLOWING, FEED_POTATOES, NEW_FEED_POTATOES, FOLLOWING};

    private static final String DELETE = "*DELETE*";

    LocalDatabaseHandler(Context context) {
        super(context, EPOTATO, null, 1);
    }

    private static String escape(String s) {
        return s != null ? s.replaceAll("\'", "\'\'") : null;
    }

    @SuppressWarnings("unused")
    @SafeVarargs
    static <T> T[] concat(T[] first, T[]... rest) {
        int totalLength = first.length;
        for (T[] array : rest)
            totalLength += array.length;
        T[] result = Arrays.copyOf(first, totalLength);
        int offset = first.length;
        for (T[] array : rest) {
            System.arraycopy(array, 0, result, offset, array.length);
            offset += array.length;
        }
        return result;
    }

    static <K, V extends Comparable<? super V>> Map<K, V> reverseSortByValue(Map<K, V> map) {
        return sortByValue(map, true);
    }

    static <K, V extends Comparable<? super V>> Map<K, V> sortByValue(Map<K, V> map) {
        return sortByValue(map, false);
    }

    private static <K, V extends Comparable<? super V>> Map<K, V> sortByValue(Map<K, V> map, boolean reverse) {
        List<Map.Entry<K, V>> list = new LinkedList<>(map.entrySet());
        Collections.sort(list, new Comparator<Map.Entry<K, V>>() {
            @Override
            public int compare(Map.Entry<K, V> o1, Map.Entry<K, V> o2) {
                return (o1.getValue()).compareTo(o2.getValue());
            }
        });
        if (reverse)
            Collections.reverse(list);
        Map<K, V> result = new LinkedHashMap<>();
        for (Map.Entry<K, V> entry : list)
            result.put(entry.getKey(), entry.getValue());
        return result;
    }

    static String serialize(ArrayList<String> in) {
        String s = "";
        try {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            new ObjectOutputStream(out).writeObject(in);
            s = new String(Hex.encodeHex(out.toByteArray()));
        } catch (IOException e) {
            e.printStackTrace();
        }
        return s;
    }

    static ArrayList<String> deserialize(String in) {
        ArrayList<String> a = new ArrayList<>();
        if (in != null && !in.equals("null") && in.length() > 0)
            try {
                a = (ArrayList<String>) new ObjectInputStream(new ByteArrayInputStream(Hex.decodeHex(in.toCharArray()))).readObject();
            } catch (IOException | ClassNotFoundException | DecoderException e) {
                e.printStackTrace();
            }
        return a;
    }

    static String sha256(String input) {
        String result = input;
        if (input != null) {
            try {
                MessageDigest md = MessageDigest.getInstance("SHA-256");
                byte[] bytes = md.digest(input.getBytes());
                StringBuilder sb = new StringBuilder();
                for (byte aByte : bytes)
                    sb.append(Integer.toString((aByte & 0xff) + 0x100, 16).substring(1));
                result = sb.toString();
            } catch (NoSuchAlgorithmException e) {
                e.printStackTrace();
            }
        }
        return result;
    }

    static String getTimestamp() {
        return new SimpleDateFormat("yyyyMMddHHmmss", Locale.UK).format(Calendar.getInstance().getTime());
    }

    static String getNiceDate(String timestamp, Context c) throws ParseException {
        String s_today = (new SimpleDateFormat("yyyyMMddHHmmss", Locale.UK)).format(Calendar.getInstance(TimeZone.getTimeZone("UTC")).getTime());
        Date dtoday = (new SimpleDateFormat("yyyyMMddHHmmss", Locale.UK)).parse(s_today);
        Date dthen = (new SimpleDateFormat("yyyyMMddHHmmss", Locale.UK)).parse(timestamp);
        Calendar ctoday = Calendar.getInstance();
        ctoday.setTime(dtoday);
        Calendar cthen = Calendar.getInstance();
        cthen.setTime(dthen);
        if (ctoday.get(Calendar.DAY_OF_YEAR) == cthen.get(Calendar.DAY_OF_YEAR) && ctoday.get(Calendar.YEAR) == cthen.get(Calendar.YEAR)) {
            DateFormat df = DateFormat.getTimeInstance(DateFormat.SHORT, Locale.getDefault());
            return df.format(dthen);
        } else if (c != null && ctoday.get(Calendar.DAY_OF_YEAR) == (cthen.get(Calendar.DAY_OF_YEAR) + 1) && ctoday.get(Calendar.YEAR) == cthen.get(Calendar.YEAR) ||
                ctoday.get(Calendar.DAY_OF_YEAR) == 1 && cthen.get(Calendar.DAY_OF_YEAR) == cthen.getActualMaximum(Calendar.DAY_OF_YEAR) && ctoday.get(Calendar.YEAR) == cthen.get(Calendar.YEAR) + 1) {
            return c.getString(R.string.yesterday);
        } else if (cthen.get(Calendar.DAY_OF_YEAR) < ctoday.get(Calendar.DAY_OF_YEAR) && ctoday.get(Calendar.DAY_OF_YEAR) - cthen.get(Calendar.DAY_OF_YEAR) < 7 ||
                cthen.get(Calendar.DAY_OF_YEAR) > ctoday.get(Calendar.DAY_OF_YEAR) && Math.abs(ctoday.get(Calendar.DAY_OF_YEAR) - cthen.get(Calendar.DAY_OF_YEAR) + cthen.getActualMaximum(Calendar.DAY_OF_YEAR)) < 7) {
            return cthen.getDisplayName(Calendar.DAY_OF_WEEK, Calendar.LONG, Locale.getDefault());
        } else {
            DateFormat df = DateFormat.getDateInstance(DateFormat.SHORT, Locale.getDefault());
            return df.format(dthen);
        }
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL("CREATE TABLE " + CONTACTS + "(" +
                ID + " TEXT, " +
                UID + " TEXT, " +
                EPID + " TEXT, " +
                CONTACT_NAME + " TEXT, " +
                CONTACT_PHONE + " TEXT, " +
                CONTACTS_SCORE + " TEXT)");
        db.execSQL("CREATE TABLE " + RECEIVED_POTATOES + "(" +
                ID + " TEXT, " +
                UID + " TEXT, " +
                RECEIVED_POTATOES_TS + " TEXT, " +
                DT + " TEXT, " +
                POTATO_TEXT + " TEXT, " +
                POTATO_FORM + " TEXT)");
        db.execSQL("CREATE TABLE " + SENT_POTATOES + "(" +
                ID + " TEXT, " +
                SENT_POTATOES_UIDS + " TEXT, " +
                SENT_POTATOES_TSS + " TEXT, " +
                SENT_POTATOES_NAMES + " TEXT, " +
                DT + " TEXT, " +
                SENT_POTATOES_CODE + " TEXT, " +
                POTATO_TEXT + " TEXT, " +
                POTATO_FORM + " TEXT)");
        db.execSQL("CREATE TABLE " + NEW_POTATOES + "(" +
                ID + " TEXT, " +
                NEW_POTATOES_PID + " TEXT)");
        db.execSQL("CREATE TABLE " + PHONE_CONTACTS + "(" +
                ID + " TEXT, " +
                CONTACT_NAME + " TEXT, " +
                CONTACT_PHONE + " TEXT)");
        db.execSQL("CREATE TABLE " + TEMP_CONTACTS + "(" +
                ID + " TEXT, " +
                UID + " TEXT, " +
                EPID + " TEXT)");
        db.execSQL("CREATE TABLE " + SUGGESTED_CONTACTS + "(" +
                ID + " TEXT, " +
                UID + " TEXT, " +
                EPID + " TEXT, " +
                CONTACTS_SCORE + " TEXT)");
        db.execSQL("CREATE TABLE " + SUGGESTED_FOLLOWING + "(" +
                ID + " TEXT, " +
                UID + " TEXT, " +
                EPID + " TEXT, " +
                CONTACTS_SCORE + " TEXT)");
        db.execSQL("CREATE TABLE " + FEED_POTATOES + "(" +
                ID + " TEXT, " +
                UID + " TEXT, " +
                DT + " TEXT, " +
                POTATO_TEXT + " TEXT, " +
                POTATO_FORM + " TEXT)");
        db.execSQL("CREATE TABLE " + NEW_FEED_POTATOES + "(" +
                ID + " TEXT, " +
                NEW_POTATOES_PID + " TEXT)");
        db.execSQL("CREATE TABLE " + FOLLOWING + "(" +
                ID + " TEXT, " +
                UID + " TEXT, " +
                EPID + " TEXT)");
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        for (String table : TABLES)
            db.execSQL("DROP TABLE IF EXISTS " + table);
        onCreate(db);
    }

    void dropEverything() {
        for (String table : TABLES)
            truncate(table);
    }

    String get(String table, String idKey, String idValue, String column) {
        String result = null;
        String countQuery = "SELECT * FROM `" + table + "` WHERE `" + idKey + "` = '" + idValue + "'";
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery(countQuery, null);
        if (cursor.getCount() > 0 && cursor.moveToFirst() && cursor.getString(cursor.getColumnIndex(column)) != null && !cursor.getString(cursor.getColumnIndex(column)).equals("null"))
            result = cursor.getString(cursor.getColumnIndex(column));
        cursor.close();
        return result;
    }

    @SuppressWarnings("unused")
    Map<String, String> get(String table, String idKey, String idValue) {
        Map<String, String> result = new HashMap<>();
        String countQuery = "SELECT * FROM `" + table + "` WHERE `" + idKey + "` = '" + idValue + "'";
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery(countQuery, null);
        if (cursor.getCount() > 0) {
            cursor.moveToFirst();
            for (String name : cursor.getColumnNames())
                result.put(name, cursor.getString(cursor.getColumnIndex(name)) != null && !cursor.getString(cursor.getColumnIndex(name)).equals("null") ? cursor.getString(cursor.getColumnIndex(name)) : null);
        }
        cursor.close();
        return result;
    }

    private ArrayList<String> getAll(String table, String column, String idKey, String idValue) {
        ArrayList<String> result = new ArrayList<>();
        String countQuery = "SELECT * FROM `" + table + "`";
        if (idKey != null)
            countQuery = "SELECT * FROM `" + table + "` WHERE `" + idKey + "` = '" + idValue + "'";
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery(countQuery, null);
        if (cursor.getCount() > 0) {
            cursor.moveToFirst();
            result.add(cursor.getString(cursor.getColumnIndex(column)) != null && !cursor.getString(cursor.getColumnIndex(column)).equals("null") ? cursor.getString(cursor.getColumnIndex(column)) : null);
            while (cursor.moveToNext())
                result.add(cursor.getString(cursor.getColumnIndex(column)) != null && !cursor.getString(cursor.getColumnIndex(column)).equals("null") ? cursor.getString(cursor.getColumnIndex(column)) : null);
        }
        cursor.close();
        return result;
    }

    ArrayList<String> getAll(String table, String column) {
        return getAll(table, column, null, null);
    }

    private ArrayList<Map<String, String>> getAll(String table, String idKey, String idValue) {
        ArrayList<Map<String, String>> result = new ArrayList<>();
        String countQuery = "SELECT * FROM `" + table + "`";
        if (idKey != null)
            countQuery = "SELECT * FROM `" + table + "` WHERE `" + idKey + "` = '" + idValue + "'";
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery(countQuery, null);
        if (cursor.getCount() > 0) {
            cursor.moveToFirst();
            Map<String, String> row0 = new HashMap<>();
            for (String name : cursor.getColumnNames())
                row0.put(name, cursor.getString(cursor.getColumnIndex(name)) != null && !cursor.getString(cursor.getColumnIndex(name)).equals("null") ? cursor.getString(cursor.getColumnIndex(name)) : null);
            result.add(row0);
            while (cursor.moveToNext()) {
                Map<String, String> row = new HashMap<>();
                for (String name : cursor.getColumnNames())
                    row.put(name, cursor.getString(cursor.getColumnIndex(name)) != null && !cursor.getString(cursor.getColumnIndex(name)).equals("null") ? cursor.getString(cursor.getColumnIndex(name)) : null);
                result.add(row);
            }
        }
        cursor.close();
        return result;
    }

    ArrayList<Map<String, String>> getAll(String table) {
        return getAll(table, null, null);
    }

    /**
     * pairs[0] = idKey
     * pairs[1] = idValue
     *
     * @return ID, if insert
     */
    @Nullable
    private String inup(boolean insert, boolean _update, String table, String... pairs) {
        String ret_id = null;
        SQLiteDatabase db = getWritableDatabase();
        for (int i = 0; i < pairs.length; i++)
            pairs[i] = escape(pairs[i]);
        boolean delete = pairs[0].equals(DELETE) || (pairs.length > 2 && pairs[2].equals(DELETE));
        boolean update = !delete && !pairs[0].equals("") && get(table, pairs[0], pairs[1], pairs[0]) != null && !insert;
        String sql = "";
        if (delete)
            sql = "DELETE FROM `" + table + (pairs[0].equals(DELETE) ? "`" : "` WHERE `" + pairs[0] + "` = '" + pairs[1] + "'");
        else if (update) {
            String set = "";
            for (int i = 2; i < pairs.length; i++) {
                if (i % 2 == 0) {
                    if (i != 2)
                        set += ", ";
                    set += "`" + pairs[i] + "`";
                } else
                    set += " = '" + pairs[i] + "'";
            }
            if (pairs.length > 2)
                sql = "UPDATE `" + table + "` SET " + set + " WHERE `" + pairs[0] + "` = '" + pairs[1] + "'";
            else
                sql = "";
        } else if (!_update) {
            String names = "`" + ID + "`";
            String values;
            int start = 0;
            if (!pairs[0].equals(ID)) {
                String B64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
                String id;
                do {
                    id = "";
                    for (int i = 0; i < 8; i++)
                        id += B64.charAt((int) (Math.random() * 64));
                } while (getAll(table, ID).contains(id));
                values = "'" + id + "'";
                ret_id = id;
            } else {
                values = "'" + pairs[1] + "'";
                start = 2;
                ret_id = pairs[1];
            }
            for (int i = start; i < pairs.length; i++) {
                if (i % 2 == 0)
                    names += ", `" + pairs[i] + "`";
                else
                    values += ", '" + pairs[i] + "'";
            }
            sql = "INSERT INTO `" + table + "` (" + names + ")VALUES(" + values + ")";
        }
        if (!sql.equals("")) db.execSQL(sql);
        return ret_id;
    }

    String inup(String table, String... pairs) {
        return inup(false, false, table, pairs);
    }

    String insert(String table, String... pairs) {
        return inup(true, false, table, pairs);
    }

    void update(String table, String... pairs) {
        inup(false, true, table, pairs);
    }

    void delete(String table, String idKey, String idValue) {
        inup(false, false, table, idKey, idValue, DELETE);
    }

    void truncate(String table) {
        execSQL("DELETE FROM `" + table + "`");
        execSQL("VACUUM");
    }

    private void execSQL(String sql) {
        SQLiteDatabase db = getWritableDatabase();
        db.execSQL(sql);
    }
}